### PR TITLE
Add initial support for signature help

### LINKED ISSUE #200: Add signature help to the `ruby-lsp`
If #199 works and is successful in providing decent results for finding possible method definitions, we can use the same mechanism to provide signature help. This request shows the method signature for the invocation at the given cursor position, displaying arguments.

Note: there may be conflicts between possible methods. It's likely that we'll only be able to provide signature help if only a single result is found in the index.

Documentation: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_signatureHelp

### PR DESCRIPTION
**Pull request by vinistock**
### Motivation

Closes #200

We're currently inserting a method's parameters as part of completion, but that's not how the specification prescribes handling parameters.

Completion should only insert the method name and then signature help shows the parameter information while the user types.

This PR removes the parameters from completion and adds initial support for signature help.

### Implementation

The idea is that we discover which method the user is invoking and then display the information.

The main source of difficulty implementing this for Ruby are keyword and optional parameters, because it means that we can't trust the order of arguments to define the `active_parameter`.

For example

```ruby
def foo(a, b = 123, c:, d:)
end

foo(a, c:,) # => current active parameter is `d` because we skipped `b`
foo(a, d:) # => current parameter is `c` because keyword parameters don't require an order
```

Currently, we are simply not handling this complexity so that we can start simple. We'll need to design some mechanism to understand what the current active parameter is.

### Automated Tests

Added tests - including the cases where we are not handling parameters correctly yet.

### CODE REVIEW

**Comment by andyw8** [comment]
Note: You need to add `"signatureHelp": true` to the enabledFeatures list in `settings.json` to enable.

**Comment by vinistock** [comment]
I hadn't caught those! I'll see how we can improve these cases.

> Typing the arguments quickly can result in the highlighted argument being wrong.

This one may not be easy to fix. It might just be the delay from the LSP processing multiple requests for each keypress. Does it just lag?

File: lib/ruby_lsp/requests/completion.rb [hunk 1]
```diff
@@ -125,7 +125,7 @@ def on_constant_path_node_enter(node)
 
       sig { params(node: Prism::CallNode).void }
       def on_call_node_enter(node)
-        return if DependencyDetector.instance.typechecker
+        # return if DependencyDetector.instance.typechecker
```
[the remark this answers is not in the document]
Comment (vinistock) [reply lib/ruby_lsp/requests/completion.rb]: Yes, oops.

File: lib/ruby_indexer/lib/ruby_indexer/entry.rb [hunk 2]
```diff
@@ -84,6 +84,8 @@ class Parameter
       sig { returns(Symbol) }
       attr_reader :name
 
+      alias_method :decorated_name, :name
```
Comment (Morriar) [inline lib/ruby_indexer/lib/ruby_indexer/entry.rb:87]: We should add a comment explaining the difference with `name` and when one should be used or the other.

File: lib/ruby_lsp/executor.rb [hunk 3]
```diff
@@ -692,6 +727,12 @@ def initialize_request(options)
         )
       end
 
+      signature_help_provider = if enabled_features["signatureHelp"]
+        Interface::SignatureHelpOptions.new(
+          trigger_characters: ["(", " ", ","],
```
Comment (Morriar) [inline lib/ruby_lsp/executor.rb:732]: Should we also trigger on `:` like if I'm typing `bar(a:` wouldn't the help disappear?

File: lib/ruby_indexer/lib/ruby_indexer/entry.rb (hunk 2 above)
Comment (vinistock) [reply lib/ruby_indexer/lib/ruby_indexer/entry.rb:89]: Added comments to both.

File: lib/ruby_lsp/executor.rb (hunk 3 above)
Comment (vinistock) [reply lib/ruby_lsp/executor.rb:732]: It's already triggered automatically. It seems editors consider `:` as part of identifiers.

File: lib/ruby_indexer/lib/ruby_indexer/entry.rb (hunk 2 above)
Comment (Morriar) [reply lib/ruby_indexer/lib/ruby_indexer/entry.rb:89]: Should the keyword parameter `foo` be decorate as `foo:`?

File: lib/ruby_lsp/executor.rb (hunk 3 above)
Comment (Morriar) [reply lib/ruby_lsp/executor.rb:732]: Should we add a comment about this for future us?

**Review by Morriar — APPROVED** [review]

File: lib/ruby_indexer/lib/ruby_indexer/entry.rb (hunk 2 above)
Comment (vinistock) [reply lib/ruby_indexer/lib/ruby_indexer/entry.rb:89]: Ah, yeah. Good catch.

**Pull request merged**